packages/lodestar/src/network/gossip/validation/aggregateAndProof.ts

```diff
@@ -0,0 +1,170 @@
+      ...logContext,
+    });
+    //TODO: aggregate and proof pool to wait until proper slot to replay
+    return ExtendedValidatorResult.ignore;
+  }
+  if (await db.seenAttestationCache.hasAggregateAndProof(aggregateAndProof)) {
+    logger.warn("Ignored gossip aggregate and proof", {
+      reason: "seen aggregator and target index",
+      targetEpoch: aggregate.data.target.epoch,
+      ...logContext,
+    });
+    return ExtendedValidatorResult.ignore;
+  }
+  if (!hasAttestationParticipants(aggregate)) {
+    logger.warn("Rejected gossip aggregate and proof", {
+      reason: "missing attestation participants",
+      ...logContext,
+    });
+    return ExtendedValidatorResult.reject;
+  }
+
+  if (await isAttestingToInValidBlock(db, aggregate)) {
+    logger.warn("Rejected gossip aggregate and proof", {
+      reason: "attesting to invalid block",
+      ...logContext,
+    });
+    return ExtendedValidatorResult.reject;
+  }
+
+  //TODO: check pool of aggregates if already seen (not a dos vector check)
+
+  const result = await validateAggregateAttestation(config, chain, db, logger, logContext, signedAggregateAndProof);
+
+  logger.profile("gossipAggregateAndProofValidation");
+  logger.info("Received gossip aggregate and proof passed validation", logContext);
+  return result;
+}
+
+export function isValidAggregateAndProofSlot(config: IBeaconConfig, genesisTime: number, slot: Slot): boolean {
+  const milliSecPerSlot = config.params.SECONDS_PER_SLOT * 1000;
+  const currentSlotTime = getCurrentSlot(config, genesisTime) * milliSecPerSlot;
+  return (
+    (slot + ATTESTATION_PROPAGATION_SLOT_RANGE) * milliSecPerSlot + MAXIMUM_GOSSIP_CLOCK_DISPARITY >= currentSlotTime &&
+    currentSlotTime >= slot * milliSecPerSlot - MAXIMUM_GOSSIP_CLOCK_DISPARITY
+  );
+}
+
+export function hasAttestationParticipants(attestation: Attestation): boolean {
+  return Array.from(attestation.aggregationBits).filter((bit) => !!bit).length >= 1;
+}
+
+export async function validateAggregateAttestation(
+  config: IBeaconConfig,
+  chain: IBeaconChain,
+  db: IBeaconDb,
+  logger: ILogger,
+  logContext: ReturnType<typeof getLogContext>,
+  aggregateAndProof: SignedAggregateAndProof
+): Promise<ExtendedValidatorResult> {
+  const attestation = aggregateAndProof.message.aggregate;
+  const attestationPreState = await getAttestationPreState(config, chain, db, attestation.data.target);
+  if (!attestationPreState) {
+    logger.warn("Ignored gossip aggregate and proof", {reason: "missing attestation prestate", ...logContext});
+    return ExtendedValidatorResult.ignore;
+  }
+  const {state, epochCtx} = attestationPreState;
+  //committee changes on epoch, so advance only if different epoch
+  const attEpoch = computeEpochAtSlot(config, attestation.data.slot);
+  if (attEpoch > computeEpochAtSlot(config, state.slot)) {
+    processSlots(epochCtx, state, computeStartSlotAtEpoch(config, attEpoch));
```

Comment:
since we have out-of-memory issue, we have to care about the use of `processSlots` more.

what's the plan for DOS attack? i.e. some peers can send us an attestation with `attestation.data.slot` so far away from current slot.

similar concern for `processSlots` inside `getAttestationPreState`

Reply:
Attestation pre state:
Well attestation.data.slot can only be at most current slot or is rejected.
So issue is if it's way ahead of our head slot.
In that case we check if we have checkpoint state in cache.
If we don't have it we search for target.root amongst unfinalized blocks and process him up to start slot of target.epoch. In normal conditions there won't be slot processing. Issue could occur if during non-finalization, some validators are deliberately producing attestation with head target epoch and very old non finalized block. I would wait till we see how other clients solved this issue. Simply limiting this to few slots could introduce bugs and maybe get us on wrong chain.

Referenced line:
This happens if slots in next epoch are skip slots so target epoch is still previous epoch.